x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/TransformProgressGatherer.java

```diff
@@ -0,0 +1,114 @@
+import java.util.Map;
+
+/**
+ * Utility class to gather the progress information for a given config and its cursor position
+ */
+public final class TransformProgressGatherer {
+
+    private static final Logger logger = LogManager.getLogger(TransformProgressGatherer.class);
+
+    /**
+     * This gathers the progress of the current indexer given its position.
+     *
+     * @param progressListener The listener to alert on completion. Error is returned if the config has not been set on the indexer
+     */
+    public static void getProgress(Client client,
+                                   DataFrameTransformConfig config,
+                                   Map<String, Object> cursor,
+                                   ActionListener<DataFrameTransformProgress> progressListener) {
+        MultiSearchRequestBuilder request = client.prepareMultiSearch()
+            .add(client.prepareSearch(config.getSource().getIndex())
+                .setSize(0)
+                .setTrackTotalHits(true)
+                .setQuery(config.getSource().getQueryConfig().getQuery())
+                .request());
+
+        if (cursor != null) {
+            request.add(client.prepareSearch(config.getSource().getIndex())
+                .setSize(0)
+                .setTrackTotalHits(true)
+                .setQuery(docsLeftQuery(config, cursor))
+                .request());
+        }
+        ActionListener<MultiSearchResponse> multiSearchResponseActionListener = ActionListener.wrap(
+            mSearchResponse -> {
+                long[] counts = new long[]{0, 0};
+                for (int i = 0; i < mSearchResponse.getResponses().length; i++) {
+                    MultiSearchResponse.Item item = mSearchResponse.getResponses()[i];
+                    if (item.isFailure()) {
+                        progressListener.onFailure(new ElasticsearchException("Failed gather transform task progress", item.getFailure()));
+                        return;
+                    }
+                    counts[i] = item.getResponse().getHits().getTotalHits().value;
+                }
+                // It stands to reason that total document count will be greater than the remaining document count as the
+                // remaining document count query contains the user provided query as a filter, thus should at most have the same
+                // amount of documents as the total document count
+                // If the mSearchResponse only contains 1 response, that response is the total document count
+                long totalDocs = Math.max(counts[0], counts[1]);
+                long remainingDocs = mSearchResponse.getResponses().length == 1 ? totalDocs : Math.min(counts[0], counts[1]);
+                progressListener.onResponse(new DataFrameTransformProgress(totalDocs, remainingDocs));
+            },
+            progressListener::onFailure
+        );
+        ClientHelper.executeAsyncWithOrigin(client.threadPool().getThreadContext(),
```

Comment:
I think we need to execute this with headers, if the index pattern matches indexes that the user does not have access to there can be a mismatch between the query executed as part of the transform (filtered) and this query (unfiltered).

Reply:
@hendrikmuhs unauthorized indices are automatically filtered in the search? I did not know this...I thought it would just error. 

I will execute with headers for sure for uniformity and accuracy.

Reply:
@benwtrent it errors if you try to access them by the exact name, above I mean things like "data-*" which dependent on the rights can resolve to a different set of indexes.